File: graph/bfs.py
from node import Node
from collections import deque
import graph

def bfs(node, visit=None):
    if visit is None:
        visit = print
    queue = deque([node])
    visited = {}
    while queue:
        v = queue.popleft()
        visited[v] = True
        visit(v)
        for vertex in v.edges:
            if vertex not in visited and vertex not in queue:
                queue.append(vertex)

def bfs_2(s):
    level = {s : 0}
    parent = {s : None}
    i = 1
    frontier = [s]
    while frontier:
        next = []
        for u in frontier:
            for v in u.edges:
                if v not in level:
                    level[v] = i
                    parent[v] = u
                    next.append(v)
        frontier = next
        i += 1
    return (level, parent)

if __name__ == '__main__':
    g = graph.make_graph([('y','z'), ('a','b'), ('a', 'd'),  ('b', 'c'), ('c', 'd'), ('b', 'y'), ('c', 'a'), ('d', 'y'), ])
    bfs(g.vertices['a'])
    g = graph.make_graph([('y','z'), ('a','b'), ('a', 'd'),  ('b', 'c'), ('c', 'd'), ('b', 'y'), ('c', 'a'), ('d', 'y'), ])
    level , parent = bfs_2(g.vertices['a'])
    print(level)
    print(parent)

File: maxflow/Bfs.java
import java.util.HashSet;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.Queue;
import java.util.NoSuchElementException;

class Bfs implements Iterator<Vertex> {
  private Vertex currentVertex;
  private Queue<Vertex> queue;
  private Graph graph;
  private HashSet<Vertex> visited;

  public Bfs(Graph g, Vertex source) {
    graph = g;
    currentVertex = source;
    visited = new HashSet<Vertex>();
    queue = new LinkedList<Vertex>();
    queue.offer(source);
  }

  public boolean hasNext() {
    return !queue.isEmpty();
  }
  
  public Vertex next() throws NoSuchElementException {
    Vertex head =  queue.poll();
    EdgeNode edge;
    Vertex from, to;
    for(int i = 0; i < head.edges.size(); i++) {
      edge = head.edges.get(i);
      from = graph.vertices.get(edge.from);
      to = graph.vertices.get(edge.to);

      if (head == from) {
        if (!visited.contains(to)) {
          queue.offer(to);
        }
      }
      else {
        if (!visited.contains(from)) {
          queue.offer(from);
        }
      }
    }
    visited.add(head);
    return head;
  }
  
  public void remove() {}
}
